src/librustc_codegen_llvm/mir/block.rs

```diff
@@ -462,6 +462,55 @@ impl FunctionCx<'a, 'll, 'tcx> {
                     return;
                 }
 
+                if (intrinsic == Some("init") || intrinsic == Some("uninit")) &&
+                    bx.cx.layout_of(sig.output()).abi.is_uninhabited()
+                {
+                    let loc = bx.sess().codemap().lookup_char_pos(span.lo());
+                    let filename = Symbol::intern(&loc.file.name.to_string()).as_str();
+                    let filename = C_str_slice(bx.cx, filename);
+                    let line = C_u32(bx.cx, loc.line as u32);
+                    let col = C_u32(bx.cx, loc.col.to_usize() as u32 + 1);
+                    let align = tcx.data_layout.aggregate_align
+                        .max(tcx.data_layout.i32_align)
+                        .max(tcx.data_layout.pointer_align);
+
+                    let str = if intrinsic == Some("init") {
+                        "Attempted to instantiate an uninhabited type (e.g. `!`) \
+                         using mem::zeroed()"
+                    } else {
+                        "Attempted to instantiate an uninhabited type (e.g. `!`) \
+                         using mem::uninitialized()"
+                    };
```

Comment:
You can actually use `format!("... uninhabited type `{}` using ...", sig.output())` here, heh.